source/rust_verify/tests/strings.rs

```diff
@@ -267,3 +266,156 @@ test_verify_one_file! {
+            proof {
+                reveal_strlit("Hello");
+            }
+            let x0: char = x.get_char(0);
+            assert(x0 == 'H');
+        }
+        fn test_get_unicode_non_ascii_passes() {
+            let emoji_with_str = new_strlit("💩");
+            proof {
+                reveal_strlit("💩");
+            }
+            let p = emoji_with_str.get_char(0);
+            assert(p == '💩');
+        }
+        fn test_get_unicode_non_ascii_passes1() {
+            let emoji_with_str = new_strlit("abcdef💩");
+            proof {
+                reveal_strlit("abcdef💩");
+            }
+            let p = emoji_with_str.get_char(0);
+            assert(p == 'a');
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_unicode_substring_passes verus_code! {
+        use pervasive::string::*;
+        fn test_substring_passes() {
+            proof {
+                reveal_strlit("01234💩");
+                reveal_strlit("012");
+                reveal_strlit("34💩");
+            }
+            let x = new_strlit("01234💩");
+            assert(x@.len() == 6);
+
+            let x0 = x.substring_char(0,3);
+            assert(x0@.ext_equal(new_strlit("012")@));
+
+            let x1 = x.substring_char(3,6);
+            assert(x1@.ext_equal(new_strlit("34💩")@));
+
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_unicode_mixed_chars verus_code! {
+        use pervasive::string::*;
+        proof fn test() {
+            let a = new_strlit("è ❤️");
+            reveal_strlit("è ❤️");
+            assert(a@[0] === 'è');
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_string_2_pass verus_code! {
+        use pervasive::string::*;
+        fn test() {
+            let a = String::from_str(new_strlit("ABC"));
+            reveal_strlit("ABC");
+            let b = a.as_str().substring_ascii(1, 2);
+            reveal_strlit("B");
+            assert(b@.ext_equal(new_strlit("B")@));
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_string_2_fail verus_code! {
+        use pervasive::string::*;
+        fn test() {
+            let a = String::from_str(new_strlit("ABC"));
+            reveal_strlit("ABC");
+            let b = a.as_str().substring_ascii(2, 3);
+            reveal_strlit("B");
+            reveal_strlit("C");
+            assert(b@.ext_equal(new_strlit("C")@));
+            assert(b@ === new_strlit("B")@); // FAILS
+        }
+    } => Err(e) => assert_one_fails(e)
+}
+
+test_verify_one_file! {
+    #[test] test_string_is_ascii_roundtrip verus_code! {
+        use pervasive::string::*;
+        fn test() {
+            let a = new_strlit("ABC");
+            let b = a.to_string();
+            let c = b.as_str();
+            reveal_strlit("ABC");
+            assert(a@.ext_equal(c@));
+            assert(a.is_ascii());
+            // TODO assert(b.is_ascii());
+            // TODO assert(c.is_ascii());
```

Comment:
I believe these should verify. I think some of the pervasive::string functions need to maintain/forward information about `is_ascii`.